tool/add_to_firebase.py

```diff
@@ -0,0 +1,114 @@
+# Temporary tool that reads from JSON and uploads it to Firebase
+# Relies on an undocumented .collections() API call and a hard coded list of top level collections
+
+import firebase_admin
+from firebase_admin import credentials
+from firebase_admin import firestore
+import hashlib
+from core_data_modules.logging import Logger
+import time
+
+import json
+import sys
+
+log = Logger(__name__)
+
+firebase_client = None
+
+def init(CRYPTO_TOKEN_PATH):
+    global firebase_client
+    log.info("Setting up Firebase client")
+    firebase_cred = credentials.Certificate(CRYPTO_TOKEN_PATH)
+    firebase_admin.initialize_app(firebase_cred)
+    firebase_client = firestore.client()
+    log.info("Done")
+
+def push_collection_to_firestore(collection_root, documents):
+    log.info(f"push_collection_to_firestore {collection_root}")
+    col = firebase_client.collection(collection_root)
+    time_start = time.perf_counter_ns()
+    for document_dict in documents:
+        push_document_to_firestore(collection_root, document_dict)
+    time_end = time.perf_counter_ns()
+    ms_elapsed = (time_end - time_start) / (1000 * 1000)
+    log.info(f"push_collection_to_firestore {collection_root} in {ms_elapsed} ms")
+
+def push_document_to_firestore(collection_root, data):
+    doc_id = list(data.keys())[0]
+    doc_fields = data[doc_id]
+    ref_path = f"{collection_root}/{doc_id}"
+    firebase_client.document(ref_path).set(doc_fields)
+
+def text_to_doc_id(prefix, text):
+    return f"{prefix}-" + hashlib.sha256(text.encode("utf-8")).hexdigest()[0:8]
+
+if (len(sys.argv) != 4):
+    print ("Usage python add_to_firebase.py crypto_token suggested_replies|message_tags|conversation_tags input_path")
+    print ("Adds data items that don't currently exist, and overwrittes items that already exist.")
+    exit(1)
+
+CRYPTO_TOKEN_PATH = sys.argv[1]
+CONTENT_TYPE = sys.argv[2]
+INPUT_PATH = sys.argv[3]
+init(CRYPTO_TOKEN_PATH)
+
+valid_content_type = ["suggested_replies", "message_tags", "conversation_tags"]
+if CONTENT_TYPE not in valid_content_type:
+    log.error("Content type '{}' not known, expected one of {}".format(CONTENT_TYPE, valid_content_type))
+    exit(1)
+
+with open(INPUT_PATH, 'r') as f:
+    data_dict = json.load(f)
+
+if CONTENT_TYPE == "suggested_replies":
+    # Make sure that the fields that the Firestore model expects exist
+    assert len(data_dict.keys()) == 1
```

Comment:
Move to validator in the fullness of time

Reply:
Thanks, I've added a couple of comments to keep track of this and created an issue too.